File: 874-BackspaceStringCompare/874-BackspaceStringCompare.py
# Last updated: 8/5/2025, 2:56:48 PM
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        p1, p2 = len(s) - 1, len(t) - 1

        while p1 >= 0 or p2 >= 0:
            if p1 >= 0 and s[p1] == "#" or p2 >= 0 and t[p2] == "#":
                if p1 >= 0 and s[p1] == "#":
                    back_count = 2
                    while back_count > 0 and p1 >= 0:
                        p1 -= 1
                        back_count -= 1
                        if p1 >= 0 and s[p1] == "#":
                            back_count += 2

                if p2 >= 0 and t[p2] == "#":
                    back_count = 2
                    while back_count > 0 and p2 >= 0:
                        p2 -= 1
                        back_count -= 1
                        if p2 >= 0 and t[p2] == "#":
                            back_count += 2
            else:
                if p1 >= 0 and p2 >= 0 and s[p1] != t[p2]:
                    return False
                if (p1 >= 0) != (p2 >= 0):  # One is exhausted before the other
                    return False

                p1 -= 1
                p2 -= 1

        return True

File: 874-BackspaceStringCompare/874-BackspaceStringCompare.java
// Last updated: 8/5/2025, 2:56:50 PM
class Solution {
    
    public boolean backspaceCompare(String S, String T) {
        return getString(S).equals(getString(T));
    }
    
    private String getString(String str) {
        int n = str.length(), count = 0;
        StringBuilder sb = new StringBuilder();
        for(int i=n-1; i>=0; i--) {
            char c = str.charAt(i);
            if(c == '#')
                count++;
            else {
                if(count > 0)
                    count--;
                else
                    sb.append(c);
            }
        }
        return sb.toString();
    }
    
}